tests/api/test_ratelimiting.py

```diff
@@ -314,3 +314,77 @@ def consume_at(time: float) -> bool:
 
         # Check that we get rate limited after using that token.
         self.assertFalse(consume_at(11.1))
+
+    def test_record_action_which_doesnt_fill_bucket(self) -> None:
+        limiter = Ratelimiter(
+            store=self.hs.get_datastores().main, clock=None, rate_hz=0.1, burst_count=3
+        )
+
+        # Observe two actions, leaving room in the bucket for one more.
+        limiter.record_action(requester=None, key="a", n_actions=2, _time_now_s=0.0)
+
+        # We should be able to take a new action now.
+        success, _ = self.get_success_or_raise(
+            limiter.can_do_action(requester=None, key="a", _time_now_s=0.0)
+        )
```

Comment:
I'm struggling a little to understand how this differs from `get_success(...)` but it seems consistent with the other tests.

Reply:
They look remarkably similar. AFAICS the only nontrivial difference is:

- `get_success_or_raise` raises an exception with Failure.raiseException() if the deferred contains a Failure.
- `get_success` (via `successResultOf`) will immediately fail the test.

Assuming that's correct, the two functions are essentially the same unless you're within a `try` block.

Reply:
Thanks! I figured it was some nuance like that but couldn't grok it immediately.

Reply:
Definitely wasn't obvious to me!